From ee8f2f74b34734b95d2a0804e12a540ed874c457 Mon Sep 17 00:00:00 2001
From: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Date: Fri, 1 Feb 2019 20:18:35 +0530
Subject: [PATCH 0146/1851] ASoC: xlnx: add multi sample rate support for SDI
 audio

commit 835e8aac2a863c8661981a7c4d65dafeff69a3fb from
https://github.com/Xilinx/linux-xlnx.git

Added mulitple sample rate support for SDI audio.

Signed-off-by: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 sound/soc/xilinx/xlnx_pl_snd_card.c | 18 +++---------------
 1 file changed, 3 insertions(+), 15 deletions(-)

diff --git a/sound/soc/xilinx/xlnx_pl_snd_card.c b/sound/soc/xilinx/xlnx_pl_snd_card.c
index 47177a76bd6b..ab54827a677c 100644
--- a/sound/soc/xilinx/xlnx_pl_snd_card.c
+++ b/sound/soc/xilinx/xlnx_pl_snd_card.c
@@ -71,21 +71,8 @@ static int xlnx_sdi_card_hw_params(struct snd_pcm_substream *substream,
 	struct pl_card_data *prv = snd_soc_card_get_drvdata(rtd->card);
 	u32 sample_rate = params_rate(params);
 
-	switch (sample_rate) {
-	case 32000:
-		prv->mclk_ratio = 576;
-		break;
-	case 44100:
-		prv->mclk_ratio = 418;
-		break;
-	case 48000:
-		prv->mclk_ratio = 384;
-		break;
-	default:
-		return -EINVAL;
-	}
-
-	return 0;
+	prv->mclk_val = prv->mclk_ratio * sample_rate;
+	return clk_set_rate(prv->mclk, prv->mclk_val);
 }
 
 static int xlnx_hdmi_card_hw_params(struct snd_pcm_substream *substream,
@@ -252,6 +239,7 @@ static struct snd_soc_dai_link xlnx_snd_dai[][XLNX_MAX_PATHS] = {
 		{
 			.name = "xlnx-sdi-playback",
 			SND_SOC_DAILINK_REG(xlnx_sdi_tx),
+			.ops = &xlnx_sdi_card_ops,
 		},
 		{
 			.name = "xlnx-sdi-capture",
-- 
2.31.1

